inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/InstrumentationManager.java

```diff
@@ -0,0 +1,209 @@
+package rocks.inspectit.oce.core.instrumentation;
+
+import com.google.common.cache.CacheBuilder;
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+import net.bytebuddy.description.type.TypeDescription;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.event.EventListener;
+import org.springframework.stereotype.Service;
+import rocks.inspectit.oce.core.config.InspectitConfigChangedEvent;
+import rocks.inspectit.oce.core.config.InspectitEnvironment;
+import rocks.inspectit.oce.core.config.model.instrumentation.InternalSettings;
+import rocks.inspectit.oce.core.instrumentation.config.ClassInstrumentationConfiguration;
+import rocks.inspectit.oce.core.instrumentation.config.InstrumentationConfigurationChangedEvent;
+import rocks.inspectit.oce.core.instrumentation.config.InstrumentationConfigurationResolver;
+import rocks.inspectit.oce.core.service.BatchJobExecutorService;
+import rocks.inspectit.oce.core.utils.StopWatch;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import java.lang.instrument.Instrumentation;
+import java.time.Duration;
+import java.util.*;
+
+/**
+ * This class is responsible for making sure that for every class the instrumentation
+ * eventually is applied as specified in the active {@link rocks.inspectit.oce.core.config.model.InspectitConfig}
+ * <p>
+ * Whenever a classes instrumentation configuration might have changes, this class gets put into this services working queue.
+ * This service is then responsible for filtering out the classes whose instrumentaiton actually has
+ * changed and triggers the retransform for these.
+ */
+@Service
+@Slf4j
+public class InstrumentationManager implements IClassDiscoveryListener {
+
+    @Autowired
+    private BatchJobExecutorService executor;
+
+    @Autowired
+    private InspectitEnvironment env;
+
+    @Autowired
+    private NewClassDiscoveryService classDisoveryService;
```

Comment:
I am totally confused now, how is this needed here when this class implements the `IClassDiscoveryListener`, how is the context actually starting as this seems to be a cyclic dependency?